File: backend/tournament/sms/adapters/ovh.py
import hashlib
import time
import logging
import requests
from ..base import BaseSmsAdapter

logger = logging.getLogger('tournament.sms')


class OvhAdapter(BaseSmsAdapter):
    """Adaptateur OVH SMS."""

    def send(self, to: str, message: str, sender: str = '') -> dict:
        app_key = self.config.get('application_key', '')
        app_secret = self.config.get('application_secret', '')
        consumer_key = self.config.get('consumer_key', '')
        service_name = self.config.get('service_name', '')
        sender_name = sender or self.config.get('sender_name', '')

        url = f"https://eu.api.ovh.com/1.0/sms/{service_name}/jobs"
        body = {
            'message': message,
            'receivers': [to],
            'noStopClause': True,
            'priority': 'high',
            'charset': 'UTF-8',
        }
        if sender_name:
            body['sender'] = sender_name

        timestamp = str(int(time.time()))
        method = 'POST'
        import json
        body_str = json.dumps(body)
        pre_hash = f"{app_secret}+{consumer_key}+{method}+{url}+{body_str}+{timestamp}"
        signature = '$1$' + hashlib.sha1(pre_hash.encode('utf-8')).hexdigest()

        headers = {
            'Content-Type': 'application/json',
            'X-Ovh-Application': app_key,
            'X-Ovh-Consumer': consumer_key,
            'X-Ovh-Timestamp': timestamp,
            'X-Ovh-Signature': signature,
        }

        try:
            resp = requests.post(url, json=body, headers=headers, timeout=15)
            data = resp.json()
            if resp.status_code == 200:
                ids = data.get('ids', [])
                return {'success': True, 'message_id': str(ids[0]) if ids else '', 'error': ''}
            else:
                error = data.get('message', str(resp.status_code))
                logger.error(f"[SMS OVH] Erreur: {error}")
                return {'success': False, 'message_id': '', 'error': error}
        except Exception as e:
            logger.error(f"[SMS OVH] Exception: {e}")
            return {'success': False, 'message_id': '', 'error': str(e)}

    @classmethod
    def get_required_config_fields(cls) -> list:
        return [
            {'name': 'application_key', 'label': 'Application Key', 'type': 'text', 'required': True, 'help_text': 'Cle d\'application OVH'},
            {'name': 'application_secret', 'label': 'Application Secret', 'type': 'password', 'required': True, 'help_text': 'Secret d\'application OVH'},
            {'name': 'consumer_key', 'label': 'Consumer Key', 'type': 'password', 'required': True, 'help_text': 'Cle consommateur OVH'},
            {'name': 'service_name', 'label': 'Nom du service', 'type': 'text', 'required': True, 'help_text': 'Nom du service SMS OVH (ex: sms-xx12345-1)'},
            {'name': 'sender_name', 'label': 'Expediteur', 'type': 'text', 'required': False, 'help_text': 'Nom expediteur (optionnel)'},
        ]

File: packages/sms/src/adapters/ovh.ts
/**
 * Adaptateur OVH SMS Pro — ACTIF en v1.
 *
 * Référence officielle : https://docs.ovh.com/fr/sms/
 * Port TypeScript du `backend/tournament/sms/adapters/ovh.py` (dépôt A).
 *
 * Authentification par signature SHA1 avec préfixe '$1$' :
 *   signature = '$1$' + sha1(appSecret + '+' + consumerKey + '+' + method + '+' + url + '+' + body + '+' + timestamp)
 */

import crypto from 'node:crypto';
import { BaseSmsAdapter } from '../adapter';
import type { AdapterField, SmsSendResult } from '@tt/types';

const OVH_API_BASE = 'https://eu.api.ovh.com/1.0';

interface OvhConfig {
  appKey: string;
  appSecret: string;
  consumerKey: string;
  serviceName: string;
  senderName?: string;
}

export class OvhSmsAdapter extends BaseSmsAdapter {
  readonly type = 'ovh' as const;
  readonly name = 'OVH SMS Pro';

  override requiredFields(): AdapterField[] {
    return [
      {
        name: 'appKey',
        label: 'Application Key',
        type: 'text',
        required: true,
        help: "Clé d'application OVH (https://api.ovh.com/createToken)",
      },
      {
        name: 'appSecret',
        label: 'Application Secret',
        type: 'password',
        secret: true,
        required: true,
      },
      {
        name: 'consumerKey',
        label: 'Consumer Key',
        type: 'password',
        secret: true,
        required: true,
      },
      {
        name: 'serviceName',
        label: 'Service Name',
        type: 'text',
        required: true,
        placeholder: 'sms-xxxxx-1',
        help: "Identifiant du service SMS OVH (ex : sms-ab12345-1)",
      },
      {
        name: 'senderName',
        label: 'Expéditeur (optionnel)',
        type: 'text',
        required: false,
        placeholder: 'ChellesTT',
        help: '11 caractères max alphanumériques. Si vide, utilise le sender par défaut.',
      },
    ];
  }

  override async send(to: string, message: string, sender?: string): Promise<SmsSendResult> {
    const cfg = this.config as Partial<OvhConfig>;
    if (!cfg.appKey || !cfg.appSecret || !cfg.consumerKey || !cfg.serviceName) {
      return { success: false, error: 'OVH non configuré (appKey/appSecret/consumerKey/serviceName manquants)' };
    }

    const url = `${OVH_API_BASE}/sms/${cfg.serviceName}/jobs`;
    const senderName = sender || cfg.senderName || '';

    const payload: Record<string, unknown> = {
      message,
      receivers: [to],
      noStopClause: true,
      priority: 'high',
      charset: 'UTF-8',
    };
    if (senderName) payload.sender = senderName;

    const bodyStr = JSON.stringify(payload);
    const timestamp = Math.floor(Date.now() / 1000).toString();
    const method = 'POST';
    const preHash = `${cfg.appSecret}+${cfg.consumerKey}+${method}+${url}+${bodyStr}+${timestamp}`;
    const signature = '$1$' + crypto.createHash('sha1').update(preHash).digest('hex');

    const headers = {
      'Content-Type': 'application/json',
      'X-Ovh-Application': cfg.appKey,
      'X-Ovh-Consumer': cfg.consumerKey,
      'X-Ovh-Timestamp': timestamp,
      'X-Ovh-Signature': signature,
    };

    try {
      const res = await fetch(url, {
        method,
        headers,
        body: bodyStr,
        signal: AbortSignal.timeout(15000),
      });
      const data = (await res.json().catch(() => ({}))) as Record<string, unknown>;

      if (res.ok) {
        const ids = (data.ids as unknown[]) ?? [];
        return {
          success: true,
          providerId: ids[0] != null ? String(ids[0]) : undefined,
          rawResponse: data,
        };
      }
      return {
        success: false,
        error: typeof data.message === 'string' ? data.message : `HTTP ${res.status}`,
        rawResponse: data,
      };
    } catch (e) {
      return {
        success: false,
        error: e instanceof Error ? e.message : String(e),
      };
    }
  }
}
